### PR TITLE
The Big Refactor

### PR DESCRIPTION
**Pull request by bushrat011899**
# Objective

It has been discovered through fuzzing that the current `bevy_mikktspace` is no longer byte-compatible with the original C implementation. Using `git bisect`, I determined that the discrepancy was introduced _before_ the transition to safe Rust. This makes diagnosing and resolving the issue almost impossible.

So I started again from scratch.

Using [`c2rust`](https://github.com/immunant/c2rust), [`cargo-fuzz`](https://github.com/rust-fuzz/cargo-fuzz), test data produced by @atlv24, and the original safe translation by @LaylBongers, I recreated a safe port of `mikktspace` in safe Rust [here](github.com/bushrat011899/mikktspace-rs).

This PR moves the changes there to `bevy_mikktspace`.

## Solution

- Added fuzzing using `cargo-fuzz` with built-in comparison to `mikktspace-sys` (the original C implementation with Rust bindings).
- Added test data from atlv24 to better catch known edge cases.
- Adjusted `Geometry` API to avoid mutually exclusive trait methods for setting tangent values.
- Migrated `no_std` support to directly receive the two missing float methods (`sqrt` and `acos`) rather than a higher level `Vec3` implementation.
- Added an `std` feature to provide a byte-compatible default. This is motivated by a desire to make this a viable replacement for `mikktspace-sys` for non-Bevy use-cases.
- Massively refactored the original C implementation to more idiomatic Rust. This includes better utilization of iterator techniques rather than preallocated buffers and indexing.
- Added documentation explaining how and why certain code is employed, especially where a more idiomatic option _isn't_ used.
- Added features to provide more "correct" results (fixing original C bugs) at the expense of byte-compatibility.
- Supersedes #5

---

## Notes

This is a complete replacement of the original code, so it may be easier to just check out and read rather than reviewing online.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -15,15 +18,37 @@ license = "Zlib AND (MIT OR Apache-2.0)"
 keywords = ["bevy", "3D", "graphics", "algorithm", "tangent"]
 rust-version = "1.85.0"
 
-exclude = ["/data", "/generator"]
+exclude = ["/data", "/fuzz"]
 
 [dependencies]
-bitflags = "2.3"
```
Comment (bushrat011899) [inline Cargo.toml:21]: Because of alignment, there was actually no space savings in the Rust implementation using `bitflags` over the far simpler "just use 4 bools".

File: Cargo.toml [hunk 2]
```diff
@@ -15,15 +18,37 @@ license = "Zlib AND (MIT OR Apache-2.0)"
...
+# Corrects a sorting bug in the original C implementation.
+# See https://github.com/mmikk/MikkTSpace/issues/5 for details
+corrected-edge-sorting = []
+
+# Uses a BTreeMap to weld vertices which is guaranteed to use the smallest vertex
+# indices.
+# This handles `NaN` values differently to the C implementation, so will produce
+# different results for poor geometry.
+# For typical geometry, this should produce identical results.
+corrected-vertex-welding = []
```
Comment (bushrat011899) [inline Cargo.toml:42]: These features might be worth prefixing with "unstable" so we can hit 1.0.

File: examples/cube_tangents.rs [hunk 3]
```diff
@@ -23,31 +21,12 @@ struct Mesh {
     vertices: Vec<Vertex>,
 }
 
-struct GlamSpace;
-
-impl bevy_mikktspace::VectorSpace for GlamSpace {
```
Comment (bushrat011899) [inline examples/cube_tangents.rs:28]: I know this was something I introduced only a few weeks ago, but having spent all that time neck deep in the unsafe C version, I think I prefer the simplicity of just giving the 2 missing float methods.

Plus, an `std` feature on by default improves the typical user experience.

File: examples/cube_tangents.rs [hunk 4]
```diff
@@ -56,27 +35,32 @@ impl bevy_mikktspace::Geometry for Mesh {
...
 
     fn tex_coord(&self, face: usize, vert: usize) -> [f32; 2] {
-        vertex(self, face, vert).tex_coord.into()
+        vertex(self, face, vert).tex_coord
     }
 
-    fn set_tangent_encoded(&mut self, tangent: [f32; 4], face: usize, vert: usize) {
+    fn set_tangent(
+        &mut self,
+        tangent: Option<bevy_mikktspace::TangentSpace>,
```
Comment (bushrat011899) [inline examples/cube_tangents.rs:52]: This change allows both `set_tangent` methods to be combined, avoiding the somewhat ugly "implement one of these" pattern that was a holdover from the original C implementation.

Also note I've explicitly replaced this with an `Option`. The original C implementation will silently return a default value under certain circumstances, which I think is something worth highlighting to the user, as they may have a more sensible default. `TangentSpace` implements `Default`, so users can get the exact behaviour anyway.

File: examples/cube_tangents.rs [hunk 5]
```diff
@@ -280,7 +264,7 @@ fn make_cube() -> Mesh {
 
 fn main() {
     let mut cube = make_cube();
-    bevy_mikktspace::generate_tangents(&mut cube);
+    let _ = bevy_mikktspace::generate_tangents(&mut cube);
```
Comment (bushrat011899) [inline examples/cube_tangents.rs:267]: We don't return `bool` in Rust to indicate success! We have `Result`!

File: fuzz/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+cargo-fuzz = true
+
+[dependencies]
+libfuzzer-sys = "0.4"
+arbitrary = { version = "1", features = ["derive"] }
+
+[dependencies.bevy_mikktspace]
+path = ".."
+
+[dependencies.mikktspace-sys]
```
Comment (bushrat011899) [inline fuzz/Cargo.toml:23]: Fuzzing always compares against the C implementation. This allows us to validate our implementation is resilient to garbage data _and_ produces identical results.

File: fuzz/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+version = "0.1.1"
+
+[[bin]]
+name = "one_triangle"
```
Comment (bushrat011899) [inline fuzz/Cargo.toml:27]: Simple test that gets a single totally random triangle. It's not particularly useful for fuzzing, mostly just a sanity check.

File: fuzz/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+path = "src/one_triangle.rs"
+test = false
+doc = false
+bench = false
+
+[[bin]]
+name = "many_faces"
```
Comment (bushrat011899) [inline fuzz/Cargo.toml:34]: This checks for an arbitrary number of arbitrary sized faces with the default threshold (disabled). This is the most useful fuzzer.

File: fuzz/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+path = "src/many_faces.rs"
+test = false
+doc = false
+bench = false
+
+[[bin]]
+name = "many_faces_threshold"
```
Comment (bushrat011899) [inline fuzz/Cargo.toml:41]: This also randomises the threshold value. Not super useful, but worth including for coverage reasons.

File: fuzz/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Geometry {
+    pub vertices: Vec<Vertex>,
+    pub faces: Vec<Face>,
+}
+
+impl Geometry {
+    pub fn validate(&mut self) -> Result<(), arbitrary::Error> {
```
Comment (bushrat011899) [inline fuzz/src/lib.rs:65]: Unfortunately, our implementation and the C one will both panic/crash under a lot of circumstances. This function pre-screens the random data to ensure we get useful test cases.

File: fuzz/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+            .any(|v| !v.is_normal())
+        {
+            return Err(arbitrary::Error::IncorrectFormat);
+        }
+
+        Ok(())
+    }
+
+    /// Compares two geometries totally before panicking _if_ they differ.
+    pub fn assert_eq(&self, other: &Self) {
```
Comment (bushrat011899) [inline fuzz/src/lib.rs:140]: It's annoying to fuzz and see a binary yes/no whether it matches the C implementation. This method totally compares results so we can see exactly the extent of the discrepancy.

File: src/mikktspace/quick_sort.rs [hunk 12]
```diff
@@ -0,0 +1,8 @@
+//! Implements edge sorting using `[T]::sort`.
+
+use super::Edge;
+
+/// Correctly sorts [`Edge`]s using a built-in sorting algorithm.
+pub(super) fn quick_sort_edges(edges: &mut [Edge]) {
+    edges.sort_unstable();
+}
```
Comment (bushrat011899) [inline src/mikktspace/quick_sort.rs:8]: If only the C implementation actually implemented sorting correctly...

File: src/mikktspace/quick_sort_legacy.rs [hunk 13]
```diff
@@ -0,0 +1,144 @@
+//! This is a deliberately _broken_ implementation of quick-sort.
```
Comment (bushrat011899) [inline src/mikktspace/quick_sort_legacy.rs:1]: I've tried to fully document exactly how the original implementation differs from a proper sort. I don't like the idea of including "bad" code without fully grasping how it breaks things.

File: src/mikktspace/raw_tangent_space.rs [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+impl<O: Ops> RawTangentSpace<O> {
+    /// Combines two [`RawTangentSpace`] values using an algorithm _similar_ to
+    /// averaging.
+    /// This is not strictly a mean, as it combines vector magnitudes and directions
+    /// separately.
+    /// For example, if `lhs.s` and `rhs.t` are `a` and `-a` (for some arbitrary vector `a`),
+    /// the expected mean would be `0`.
+    /// However, this method will instead return `a` (as `(a + a) / 2 == a`).
+    pub(super) fn combine(self, rhs: Self) -> Self {
```
Comment (bushrat011899) [inline src/mikktspace/raw_tangent_space.rs:71]: I'm pretty sure the original C implementation is wrong here. Or at least, the method shouldn't have been called "average". The way it combines direction and magnitude is not an average, so I've documented it and named it `combine` to be more clear.

File: src/mikktspace/weld_vertices.rs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    }
+}
+
+/// Compares two vertices using [`f32::total_cmp`] for each component of its position,
+/// normal, and texture coordinate.
+#[derive(Clone, Copy)]
+struct Key([f32; 3 + 3 + 2]);
+
+impl Key {
+    fn new<I: MikkTSpaceInterface<O>, O: Ops>(context: &I, index: FaceVertex) -> Self {
```
Comment (bushrat011899) [inline src/mikktspace/weld_vertices.rs]: TODO: forgot to fix the trait name because this code is not compiled by default. Will amend CI to check features compile...

File: tests/sample_data.rs [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+generate_tests! {
+    cube: "../data/cube.obj",
+    crangeract: "../data/crangeract.obj",
+    doomcone_smooth: "../data/doomcone_smooth.obj",
+    doomcone: "../data/doomcone.obj",
+    obliterated: "../data/obliterated.obj",
+    rancid_geometry: "../data/rancid_geometry.obj",
+    tangeract: "../data/tangeract.obj",
+}
```
Comment (bushrat011899) [inline tests/sample_data.rs:218]: For all of these tests we directly compare to the C implementation to ensure identical results.

File: src/mikktspace.rs [hunk 17]
```diff
@@ -0,0 +1,901 @@
...
+    not(feature = "corrected-edge-sorting"),
+    path = "mikktspace/quick_sort_legacy.rs"
+)]
+mod quick_sort;
+mod raw_tangent_space;
+#[cfg_attr(
+    not(feature = "corrected-vertex-welding"),
+    path = "mikktspace/weld_vertices_legacy.rs"
+)]
+mod weld_vertices;
```
Comment (bushrat011899) [inline src/mikktspace.rs:35]: I've split out some of this file into a couple of supporting modules, just to help with clarity and feature gating.

File: src/mikktspace.rs [hunk 18]
```diff
@@ -0,0 +1,901 @@
...
+    }
+
+    Ok(())
+}
+
+/// Generate [`TangentSpace`] values for the provided [geometry](MikkTSpaceInterface).
+///
+/// This is separated from [`generate_tangent_space_and_write`] to highlight this
+/// step does not require mutable access to the provided [context](MikkTSpaceInterface).
+fn generate_tangent_space<I: Geometry<O>, O: Ops>(
```
Comment (bushrat011899) [inline src/mikktspace.rs:82]: Compared to the original C implementation, I've reordered some operations to simplify. For example, the vertex list isn't actually used until much later in the algorithm, but it's generated early and must be amended as the triangle list is modified to keep them in sync. Instead, I generate the list towards the end and skip all the intermediate busy work.

File: src/mikktspace.rs [hunk 19]
```diff
@@ -0,0 +1,901 @@
...
+    context: &I,
+    linear_threshold: f32,
+    faces_total: usize,
+) -> Result<Vec<Option<TangentSpace<O>>>, GenerateTangentSpaceError> {
+    // This set of operations can be done on-line.
+    // That is, they do not require an intermediatory buffer.
+    let mut triangle_info_list = generate_triangle_info_list(context, faces_total);
```
Comment (bushrat011899) [inline src/mikktspace.rs:89]: I've moved most of the manipulation of triangle information into this method, as the original implementation iterates over the list _many_ times, when it could instead perform most of the tasks in one go. With Rust iterators that's very trivial.

File: src/mikktspace.rs [hunk 20]
```diff
@@ -0,0 +1,901 @@
...
+impl<O: Ops> From<TangentSpace<O>> for crate::TangentSpace {
+    fn from(value: TangentSpace<O>) -> Self {
+        crate::TangentSpace::from(&value)
+    }
+}
+
+struct TriangleInfo<O: Ops> {
+    /// Stores the index of each neighboring triangle to this one, if any.
+    /// Indices correspond to _edges_ of this triangle, rather than _vertices_.
+    face_neighbors: [Option<usize>; 3],
```
Comment (bushrat011899) [inline src/mikktspace.rs:167]: `usize` increases the size of this type a lot, but it's idiomatic Rust so I'd prefer it over `u32` indexing...

**Review by bushrat011899 — COMMENTED** [review]
Just some comments to assist reviewers.

File: src/math.rs [hunk 21]
```diff
@@ -1,56 +1,205 @@
-use core::ops::{Add, Mul, Neg, Sub};
+//! # Copyright
```
Comment (alice-i-cecile) [inline src/math.rs:1]:
```suggestion
//! # Copyright
//! 
//! This code is a Rust reimplementation of https://github.com/mmikk/MikkTSpace.
//! The copyright notice below reflects that history, and should not be removed.
```

File: src/mikktspace/face_vertex.rs [hunk 22]
```diff
@@ -0,0 +1,33 @@
+/// New-type index for a particular vertex on a particular face.
```
Comment (alice-i-cecile) [inline src/mikktspace/face_vertex.rs:1]: Documenting the winding order here would be helpful.

File: src/mikktspace.rs [hunk 23]
```diff
@@ -0,0 +1,901 @@
+//! # Copyright
```
Comment (alice-i-cecile) [inline src/mikktspace.rs:1]: Same addendum / context to the copyright notice here please.

**Review by alice-i-cecile — COMMENTED** [review]
Non-malicious, completely safe, generally very clear. I am very much leaning on tests for correctness.

I have a few small nits, and I would like to clarify the copyright notices before we merge this. I *know* the standard style is "just dump it at the top of the file", but I always find that very unhelpful and unclear about why the notice exists at all.

File: src/mikktspace/face_vertex.rs (hunk 22 above)
Comment (bushrat011899) [reply src/mikktspace/face_vertex.rs:1]: The way mikktspace works _should_ be independent of the winding order:

> This is done by performing an internal welding step and subsequently an
order-independent evaluation of tangent space for meshes consisting of
triangles and quads.
This means faces can be received in any order and the same is true for the
order of vertices of each face.
The generated result will not be affected by such reordering.

But I'll leave this comment unresolved just in case someone with better rendering experience would like to weigh in!

**Comment by bushrat011899** [comment]
Added a benchmark which compares `bevy_mikktspace` against `mikktspace-sys` on a subdivided plane (something simple and can be generated on the fly to allow _huge_ models if desired). For a 32x32 quad, both implementations complete the benchmark in around 1.9ms (plus or minus 0.1ms). Changing the number of subdivisions drastically changes the runtime, but the relative performance stays largely the same; within percentage points of each other.

**Review by alice-i-cecile — APPROVED** [review]

File: src/mikktspace.rs (hunk 20 above)
Comment (atlv24) [reply src/mikktspace.rs:167]: NonMaxUsize? will this always be 0..3

Comment (bushrat011899) [reply src/mikktspace.rs:167]: Ah so this is a bit of bad documentation on my part. The `Option<usize>` value is an index into the big list of all triangles. The edge value is bounded to `0..3`, but is the index into the `face_neighbors` array itself, not the value stored in each slot. I've updated the documentation to better explain that.

As for `NonMaxUsize`, I agree, but I'd rather not. The goal is to get this to 1.0 and never worry about it again, whereas the `nonmax` crate is still pre-1.0. Technically it shouldn't matter, as it wouldn't be a part of the public API, but it'd still potentially cause duplicated dependencies in the future, etc.

**Review by atlv24 — APPROVED** [review]
I don't know how worthwhile it is to add all of the obj tests i made, id be happy with just the doomcone. Its just a lot of lines lol

These changes are great, thanks. I think the features approach to the fixes is good, and this is pretty much ready for 1.0
